### PR TITLE
Remove usages of the AccessiblePrivateMethods trait

### PR DESCRIPTION
**Pull request by Konamiman**
### Changes proposed in this Pull Request:

Deprecates the usage of the `AccessiblePrivateMethods` trait, which was introduced as a means to keep methods that exist solely as the destination of a hook as `private`, thus keeping them out of the backwards compatibility pipeline. To simplify the codebase now the preferred approach is to make the methods `public` but adding an `@internal` annotation to their documentation comment. This pull request removes all the trait usages and turns all the affected methods into `public`+`@internal`.

The trait itself will be removed in WooCommerce 10.5 so that there's enough time to refactor any possible extensions using it (which _theorically_ shouldn't be the case since the trait is in the `Internal` namespace).

The `README.md` files in `includes` and `src` are updated too to explicitly mention the meaning of the `@internal` annotation.

### How to test the changes in this Pull Request:

The gist of the review should be ensuring that for all the class that had an `use AccessiblePrivateMethod` statement:

1. The said `use` statement has been removed.
2. The `self::add_action` and `self::add_filter` statements have been converted to plain `add_action` and `add_filter`.
3. The methods that were previously the target of such `self::add_action` or `self::add_filter` and were private are now public and have got an `@internal` annotation.
4. The same for the few instances of `self::mark_method_as_accessible` and `self::mark_static_method_as_accessible` that were present (these methods aren't called, and the target methods are now `public`+`@internal`).

Note that there were a few instances of `self::add_action` and `self::add_filter` where the target method was already public, or they have a target closure instead of a target method. In these cases removing the `self::` part is enough.

Other than that, do a general smoketesting of WooCommerce to ensure that critical flows aren't broken. Test also the `wp wc palt` commands in CLI since the `Automattic\WooCommerce\Internal\ProductAttributesLookup\CLIRunner` has got its `init` method removed.

Finally, check the content added to the modified `README.md` files.

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce/includes/README.md [hunk 1]
```diff
@@ -32,3 +34,7 @@ In order to make it easier to write unit tests, there are a couple of mechanisms
 
 It's up to you as a contributor to decide which mechanism to use in each case. Choose wisely.
 
+
+## A note on `@internal` annotations
+
+Some classes, methods and functions in this folder have an `@internal` annotation in their documentation comment block. This means that the code entity is intended for internal usage of WooCommerce core only, and **must not** be used in extensions. Backwards compatibility for these code entities is not guaranteed: they could be renamed, modified (in behavior, return value or arguments accepted) or deleted.
```
Comment (coreymckrill) [inline plugins/woocommerce/includes/README.md]: I would recommend moving this higher up in the document. Maybe even right after the first blurb?

File: plugins/woocommerce/includes/class-wc-payment-gateways.php [hunk 2]
```diff
@@ -130,26 +127,30 @@ public function init() {
 		do_action( 'wc_payment_gateways_initialized', $this );
 	}
 
+	// phpcs:disable Generic.CodeAnalysis.UnusedFunctionParameter.Found
```
Comment (coreymckrill) [inline plugins/woocommerce/includes/class-wc-payment-gateways.php:130]: I think we should generally try to avoid disable statements like this. Since the function parameter isn't used, could we just rename it `$deprecated` or something and make note of it in the doc block?

File: plugins/woocommerce/src/Internal/Features/FeaturesController.php [hunk 3]
```diff
@@ -91,18 +88,18 @@ class FeaturesController {
...
-		self::add_action( 'after_plugin_row', array( $this, 'handle_plugin_list_rows' ), 10, 2 );
-		self::add_action( 'current_screen', array( $this, 'enqueue_script_to_fix_plugin_list_html' ), 10, 1 );
-		self::add_filter( 'views_plugins', array( $this, 'handle_plugins_page_views_list' ), 10, 1 );
-		self::add_filter( 'woocommerce_admin_shared_settings', array( $this, 'set_change_feature_enable_nonce' ), 20, 1 );
-		self::add_action( 'admin_init', array( $this, 'change_feature_enable_from_query_params' ), 20, 0 );
+		add_filter( 'init', array( $this, 'start_listening_for_option_changes' ), 10, 0 );
+		add_filter( 'woocommerce_get_sections_advanced', array( $this, 'add_features_section' ), 10, 1 );
+		add_filter( 'woocommerce_get_settings_advanced', array( $this, 'add_feature_settings' ), 10, 2 );
+		add_filter( 'deactivated_plugin', array( $this, 'handle_plugin_deactivation' ), 10, 1 );
+		add_filter( 'all_plugins', array( $this, 'filter_plugins_list' ), 10, 1 );
```
Comment (coreymckrill) [inline plugins/woocommerce/src/Internal/Features/FeaturesController.php:95]: This function was already public, but I wonder if we should add the `@internal` prop to the doc block anyway...

File: plugins/woocommerce/src/Internal/Features/FeaturesController.php [hunk 4]
```diff
@@ -91,18 +88,18 @@ class FeaturesController {
...
+		add_action( 'admin_notices', array( $this, 'display_notices_in_plugins_page' ), 10, 0 );
+		add_action( 'load-plugins.php', array( $this, 'maybe_invalidate_cached_plugin_data' ) );
+		add_action( 'after_plugin_row', array( $this, 'handle_plugin_list_rows' ), 10, 2 );
+		add_action( 'current_screen', array( $this, 'enqueue_script_to_fix_plugin_list_html' ), 10, 1 );
+		add_filter( 'views_plugins', array( $this, 'handle_plugins_page_views_list' ), 10, 1 );
+		add_filter( 'woocommerce_admin_shared_settings', array( $this, 'set_change_feature_enable_nonce' ), 20, 1 );
```
Comment (coreymckrill) [inline plugins/woocommerce/src/Internal/Features/FeaturesController.php:101]: This function was already public, but I wonder if we should add the `@internal` prop to the doc block anyway...

File: plugins/woocommerce/src/Internal/TransientFiles/TransientFilesEngine.php [hunk 5]
```diff
@@ -51,12 +48,12 @@ class TransientFilesEngine implements RegisterHooksInterface {
...
 	public function register() {
-		self::add_action( self::CLEANUP_ACTION_NAME, array( $this, 'handle_expired_files_cleanup_action' ) );
-		self::add_filter( 'woocommerce_debug_tools', array( $this, 'add_debug_tools_entries' ), 999, 1 );
+		add_action( self::CLEANUP_ACTION_NAME, array( $this, 'handle_expired_files_cleanup_action' ) );
+		add_filter( 'woocommerce_debug_tools', array( $this, 'add_debug_tools_entries' ), 999, 1 );
 
-		self::add_action( 'init', array( $this, 'add_endpoint' ), 0 );
-		self::add_filter( 'query_vars', array( $this, 'handle_query_vars' ), 0 );
-		self::add_action( 'parse_request', array( $this, 'handle_parse_request' ), 0 );
+		add_action( 'init', array( $this, 'add_endpoint' ), 0 );
```
Comment (coreymckrill) [inline plugins/woocommerce/src/Internal/TransientFiles/TransientFilesEngine.php:54]: Even though this method was already public, should we add the `@internal` prop to it?

**Review by coreymckrill — COMMENTED** [review]
* Reviewed all changed files, left a few minor comments for discussion
* Smoke tested front and back end
* Smoke tested `wp wc palt`

File: plugins/woocommerce/includes/README.md (hunk 1 above)
Comment (Konamiman) [reply plugins/woocommerce/includes/README.md]: Yep, good idea, the more visible the better. I did the same in the file in `src` too.

File: plugins/woocommerce/includes/class-wc-payment-gateways.php (hunk 2 above)
Comment (Konamiman) [reply plugins/woocommerce/includes/class-wc-payment-gateways.php:130]: We could, but that won't stop the code sniffer from complaining (and `@deprecated` annotations only work for whole methods).

File: plugins/woocommerce/src/Internal/Features/FeaturesController.php (hunk 3 above)
Comment (Konamiman) [reply plugins/woocommerce/src/Internal/Features/FeaturesController.php:95]: All the `@internal` annotations in the `Internal` namespace are actually unnecessary/redundant, but I added them anyway for extra visibility. Your suggestion is consistent with that approach, so yes, let's do it.

File: plugins/woocommerce/src/Internal/Admin/Settings/PaymentsRestController.php
Comment (coreymckrill) [inline plugins/woocommerce/src/Internal/Admin/Settings/PaymentsRestController.php:1]: 🤔 The changes in this file seem unrelated to AccessiblePrivateMethods... did this leak over from some other changeset?

Comment (Konamiman) [reply plugins/woocommerce/src/Internal/Admin/Settings/PaymentsRestController.php:1]: No, it's on purpose: previously, `RestApiControllerBase` included the trait and thus all the derived classes had to include it too, even if they were not using it explicitly (they were using it implicitly in the code of the base class). So now the usage of the trait must be removed from these too.

Comment (coreymckrill) [reply plugins/woocommerce/src/Internal/Admin/Settings/PaymentsRestController.php:1]: But I'm talking about the removal of all of these constants from the top of `PaymentsRestController`. Are those somehow related to the `AccessiblePrivateMethods` trait?

Comment (Konamiman) [reply plugins/woocommerce/src/Internal/Admin/Settings/PaymentsRestController.php:1]: I'm confused now... these are the only changes that GitHub is reporting for `PaymentsRestController.php` in this PR, and this seems consistent with the current state of the file in `trunk`:

image

Comment (coreymckrill) [reply plugins/woocommerce/src/Internal/Admin/Settings/PaymentsRestController.php:1]: 🤔 Oh weird, now I see a "Changed since last view" badge on this file, and the unrelated changes are now gone. Not sure what happened here, but... carry on.

**Review by coreymckrill — APPROVED** [review]
